components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/jwt/JWTValidator.java

```diff
@@ -0,0 +1,481 @@
+import java.security.NoSuchAlgorithmException;
+import java.security.InvalidKeyException;
+import java.security.cert.Certificate;
+import java.util.Base64;
+
+/**
+ * A Validator class to validate JWT tokens in an API request.
+ */
+public class JWTValidator {
+
+    private static final Log log = LogFactory.getLog(JWTValidator.class);
+
+    private String apiLevelPolicy;
+    private boolean isGatewayTokenCacheEnabled;
+
+    public JWTValidator(String apiLevelPolicy) {
+        this.apiLevelPolicy = apiLevelPolicy;
+        this.isGatewayTokenCacheEnabled = isGatewayTokenCacheEnabled();
+    }
+
+    /**
+     * Authenticates the given request with a JWT token to see if an API consumer is allowed to access
+     * a particular API or not.
+     *
+     * @param jwtToken             The JWT token sent with the API request
+     * @param synCtx               The message to be authenticated
+     * @param openAPI              The OpenAPI object of the invoked API
+     * @return an AuthenticationContext object which contains the authentication information
+     * @throws APISecurityException in case of authentication failure
+     */
+    @MethodStats
+    public AuthenticationContext authenticate(String jwtToken, MessageContext synCtx, OpenAPI openAPI)
+            throws APISecurityException {
+
+        String[] splitToken = jwtToken.split("\\.");
+
+        JSONObject payload = null;
+        boolean isVerified = false;
+
+        String tokenSignature = splitToken[2];
+
+        String apiContext = (String) synCtx.getProperty(RESTConstants.REST_API_CONTEXT);
+        String apiVersion = (String) synCtx.getProperty(RESTConstants.SYNAPSE_REST_API_VERSION);
+        String httpMethod = (String) ((Axis2MessageContext) synCtx).getAxis2MessageContext().
+                getProperty(Constants.Configuration.HTTP_METHOD);
+        String matchingResource = (String) synCtx.getProperty(APIConstants.API_ELECTED_RESOURCE);
+
+        String cacheKey = getAccessTokenCacheKey(tokenSignature, apiContext, apiVersion, matchingResource, httpMethod);
+        String tenantDomain = getTenantDomain();
+
+        // Validate from cache
+        if (isGatewayTokenCacheEnabled) {
+            String cacheToken = (String) getGatewayTokenCache().get(tokenSignature);
+            if (cacheToken != null) {
+                log.debug("Token retrieved from the token cache.");
+                isVerified = true;
+            } else if (getInvalidTokenCache().get(tokenSignature) != null) {
+                log.debug("Token retrieved from the invalid token cache.");
+                throw new APISecurityException(APISecurityConstants.API_AUTH_INVALID_CREDENTIALS,
+                        "Invalid JWT token");
+            }
+        }
+
+        // Not found in cache
+        if (!isVerified) {
+            log.debug("Token not found in the cache.");
+            try {
+                payload = new JSONObject(new String(Base64.getUrlDecoder().decode(splitToken[1])));
```

Comment:
posiible array index out of bounds/

Reply:
this split length validation is happening at the OAuthAuthenticator so when we hit this method, it should definitely have an array of length 3